From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 10 Oct 2025 14:54:21 +0900
Subject: [PATCH] Entity activation


diff --git a/src/main/java/io/papermc/paper/entity/activation/ActivationType.java b/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
index 493914c8d14a4f4b58e57ca0e2d3723e6fd4c364..a84b55d029ad3b8e83c6c8ea183ede81acd04d77 100644
--- a/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
+++ b/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
@@ -21,6 +21,7 @@ public enum ActivationType {
     RAIDER,
     MISC;
 
+    @Deprecated // Leaf - Entity activation
     AABB boundingBox = new AABB(0, 0, 0, 0, 0, 0);
 
     /**
