From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 10 Oct 2025 14:54:17 +0900
Subject: [PATCH] optimize entity activation


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 09dd84d9866a3bbb7282b3277be99a5c16cb0a74..7cc11ec11650073066d8224c0109730c9562564d 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -213,6 +213,16 @@ public final class ChunkEntitySlices {
         return collectedEntities;
     }
 
+    // Leaf start - optimize entity activation
+    public void leaf$getAllEntities(final it.unimi.dsi.fastutil.objects.ObjectArrayList<Entity> into) {
+        final int len = this.entities.size();
+        final Entity[] rawData = this.entities.getRawData();
+        if (len != 0) {
+            into.addElements(into.size(), rawData, 0, len);
+        }
+    }
+    // Leaf end - optimize entity activation
+
     public boolean isEmpty() {
         return this.entities.size() == 0;
     }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
index c1d24cff2af534edfb9fcc45f3cf2b31eb04c6de..fcdfbce42f65da08286c0b3231ecb4b45c21d796 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -633,6 +633,48 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         }
     }
 
+    // Leaf start - optimize entity activation
+    public final void leaf$getEntities(final AABB box, final it.unimi.dsi.fastutil.longs.LongOpenHashSet chunkSet, final it.unimi.dsi.fastutil.objects.ObjectArrayList<Entity> into) {
+        final int minChunkX = (Mth.floor(box.minX) - 2) >> 4;
+        final int minChunkZ = (Mth.floor(box.minZ) - 2) >> 4;
+        final int maxChunkX = (Mth.floor(box.maxX) + 2) >> 4;
+        final int maxChunkZ = (Mth.floor(box.maxZ) + 2) >> 4;
+
+        final int minRegionX = minChunkX >> REGION_SHIFT;
+        final int minRegionZ = minChunkZ >> REGION_SHIFT;
+        final int maxRegionX = maxChunkX >> REGION_SHIFT;
+        final int maxRegionZ = maxChunkZ >> REGION_SHIFT;
+
+        for (int currRegionZ = minRegionZ; currRegionZ <= maxRegionZ; ++currRegionZ) {
+            final int minZ = currRegionZ == minRegionZ ? minChunkZ & REGION_MASK : 0;
+            final int maxZ = currRegionZ == maxRegionZ ? maxChunkZ & REGION_MASK : REGION_MASK;
+
+            for (int currRegionX = minRegionX; currRegionX <= maxRegionX; ++currRegionX) {
+                final ChunkSlicesRegion region = this.getRegion(currRegionX, currRegionZ);
+
+                if (region == null) {
+                    continue;
+                }
+
+                final int minX = currRegionX == minRegionX ? minChunkX & REGION_MASK : 0;
+                final int maxX = currRegionX == maxRegionX ? maxChunkX & REGION_MASK : REGION_MASK;
+
+                for (int currZ = minZ; currZ <= maxZ; ++currZ) {
+                    for (int currX = minX; currX <= maxX; ++currX) {
+                        final ChunkEntitySlices chunk = region.get(currX | (currZ << REGION_SHIFT));
+                        if (chunk == null || !chunk.status.isOrAfter(FullChunkStatus.FULL)) {
+                            continue;
+                        }
+                        if (chunkSet.add(CoordinateUtils.getChunkKey((currRegionX << REGION_SHIFT) | currX, (currRegionZ << REGION_SHIFT) | currZ))) {
+                            chunk.leaf$getAllEntities(into);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // Leaf end - optimize entity activation
+
     public void getHardCollidingEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate) {
         final int minChunkX = (Mth.floor(box.minX) - 2) >> 4;
         final int minChunkZ = (Mth.floor(box.minZ) - 2) >> 4;
diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index b0be24fbba95f2590b1793f9e96ab19d832112de..f4f9e4ab0151b643c1e81b917442b98e8564d8ba 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -93,6 +93,7 @@ public final class ActivationRange {
     }
     // Gale end - variable entity wake-up duration
 
+    @Deprecated // Leaf - optimize entity activation
     static AABB maxBB = new AABB(0, 0, 0, 0, 0, 0);
 
     /**
@@ -154,13 +155,25 @@ public final class ActivationRange {
         maxRange = Math.max(maxRange, villagerActivationRange);
         maxRange = Math.min((world.spigotConfig.simulationDistance << 4) - 8, maxRange);
 
+        // Leaf start - optimize entity activation
+        final boolean tickMarkers = world.paperConfig().entities.markers.tick;
+        final boolean dab = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.enabled;
+        final boolean dontEnableIfInWater = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.dontEnableIfInWater;
+        final double startSq = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.startDistanceSquared;
+        final double scale = Math.pow(2.0, -org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.activationDistanceMod);
+        final int maxPriority = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.maximumActivationPrio;
+        final boolean idleTimeoutTickNearbyEntities = world.purpurConfig.idleTimeoutTickNearbyEntities;
+        final boolean ignoreSpectatorActivation = world.spigotConfig.ignoreSpectatorActivation;
+        final long currentTick = MinecraftServer.currentTick;
+        it.unimi.dsi.fastutil.ints.IntOpenHashSet reset = new it.unimi.dsi.fastutil.ints.IntOpenHashSet();
         for (final Player player : world.players()) {
-            player.activatedTick = MinecraftServer.currentTick;
-            if (world.spigotConfig.ignoreSpectatorActivation && player.isSpectator()) {
+            player.activatedTick = currentTick;
+            if (ignoreSpectatorActivation && player.isSpectator()) {
                 continue;
             }
 
-            if (!player.level().purpurConfig.idleTimeoutTickNearbyEntities && player.isAfk()) continue; // Purpur - AFK API
+            if (!idleTimeoutTickNearbyEntities && player.isAfk()) continue; // Purpur - AFK API
+        // Leaf end - optimize entity activation
 
             final int worldHeight = world.getHeight();
             ActivationRange.maxBB = player.getBoundingBox().inflate(maxRange, worldHeight, maxRange);
@@ -173,7 +186,7 @@ public final class ActivationRange {
             ActivationType.VILLAGER.boundingBox = player.getBoundingBox().inflate(villagerActivationRange, worldHeight, villagerActivationRange);
 
             final List<Entity> entities = world.getEntities((Entity) null, ActivationRange.maxBB, e -> true);
-            final boolean tickMarkers = world.paperConfig().entities.markers.tick;
+            reset.ensureCapacity(entities.size()); // Leaf - optimize entity activation
             for (final Entity entity : entities) {
                 if (!tickMarkers && entity instanceof Marker) {
                     continue;
@@ -181,21 +194,21 @@ public final class ActivationRange {
 
                 ActivationRange.activateEntity(entity);
 
+                // Leaf start - optimize entity activation
                 // Pufferfish start
-                if (org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.enabled && entity.getType().dabEnabled &&
-                        (!org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.dontEnableIfInWater || entity.getType().is(net.minecraft.tags.EntityTypeTags.CAN_BREATHE_UNDER_WATER) || !entity.isInWater())) { // Leaf - Option for dontEnableIfInWater
-                    if (!entity.activatedPriorityReset) {
-                        entity.activatedPriorityReset = true;
-                        entity.activatedPriority = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.maximumActivationPrio;
-                    }
-                    int squaredDistance = (int) player.distanceToSqr(entity);
-                    entity.activatedPriority = squaredDistance > org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.startDistanceSquared ?
-                            Math.max(1, Math.min(squaredDistance >> org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.activationDistanceMod, entity.activatedPriority)) :
-                            1;
+                final int priority;
+                if (dab && entity.getType().dabEnabled && !(dontEnableIfInWater && entity.isInWater() && !(entity instanceof net.minecraft.world.entity.animal.fish.WaterAnimal || (entity instanceof final LivingEntity livingEntity && livingEntity.canBreatheUnderwater())))) { // Leaf - Option for dontEnableIfInWater
+                    double distSq = player.distanceToSqr(entity.position);
+                    //noinspection MathClampMigration
+                    priority = distSq > startSq ?
+                        Math.min(maxPriority, Math.max((int) (distSq * scale), 1)) :
+                        1;
                 } else {
-                    entity.activatedPriority = 1;
+                    priority = 1;
                 }
+                entity.activatedPriority = reset.add(entity.getId()) ? priority : Math.min(priority, entity.activatedPriority);
                 // Pufferfish end
+                // Leaf end - optimize entity activation
             }
         }
     }
@@ -205,6 +218,7 @@ public final class ActivationRange {
      *
      * @param entity
      */
+    @Deprecated // Leaf - optimize entity activation
     private static void activateEntity(final Entity entity) {
         if (MinecraftServer.currentTick > entity.activatedTick) {
             if (entity.defaultActivationState) { // Pufferfish - diff on change
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3fb0c67855de020b219790f542d25975803d7720..c7988daec4d023c0f006413c5c939cc82abd5090 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1034,13 +1034,13 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
                 profilerFiller.pop();
             }
 
-            io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); // Paper - EAR
+            if (org.dreeam.leaf.config.modules.opt.OptimizeEntityActivation.enabled) { entityActivation.activateEntities(this); } else { io.papermc.paper.entity.activation.ActivationRange.activateEntities(this); } // Paper - EAR // Leaf - optimize entity activation
             boolean didDespawn = tickRateManager.runsNormally() && despawnMap.tick(this, this.entityTickList); // Leaf - optimize despawn
             this.globalTemptationLookup.tick(this); // Paper - optimise temptation lookups - reset global cache prior to next entity tick // Leaf - Paper PR: Optimise temptation lookups changes
             this.entityTickList
                 .forEach(
                     entity -> {
-                        entity.activatedPriorityReset = false; // Pufferfish - DAB
+                        // entity.activatedPriorityReset = false; // Pufferfish - DAB // Leaf - optimize entity activation
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
                                 profilerFiller.push("checkDespawn");
@@ -1223,6 +1223,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     public final org.dreeam.leaf.world.RandomTickSystem randomTickSystem = new org.dreeam.leaf.world.RandomTickSystem(); // Leaf - optimize random tick
     public final org.dreeam.leaf.world.EntityCollisionCache entityCollisionCache = new org.dreeam.leaf.world.EntityCollisionCache(); // Leaf - cache collision list
     public final org.dreeam.leaf.util.FastBitRadixSort fastBitRadixSort = new org.dreeam.leaf.util.FastBitRadixSort(); // Leaf - fast bit radix sort
+    public final org.dreeam.leaf.world.EntityActivation entityActivation = new org.dreeam.leaf.world.EntityActivation(); // Leaf - optimize entity activation
     public void tickChunk(LevelChunk chunk, int randomTickSpeed) {
         final net.minecraft.world.level.levelgen.BitRandomSource simpleRandom = this.simpleRandom; // Paper - optimise random ticking // Leaf - Faster random generator - upcasting
         ChunkPos pos = chunk.getPos();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 621006ced1b053c44ecc2187e03635bc188c58c3..a01d7d2cc8fe294fdd4a23562038ce205318326f 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -379,7 +379,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public final int despawnTime; // Paper - entity despawn time limit // Leaf - Rewrite entity despawn time - private -> public
     public int totalEntityAge; // Paper - age-like counter for all entities
     private int lastTickTime; // Leaf - Rewrite entity despawn time
-    public boolean activatedPriorityReset = false; // Pufferfish - DAB
+    // public boolean activatedPriorityReset = false; // Pufferfish - DAB // Leaf - optimize entity activation
     public int activatedPriority = org.dreeam.leaf.config.modules.opt.DynamicActivationofBrain.maximumActivationPrio; // Pufferfish - DAB (golf score)
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
     // Paper start - EAR 2
