From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <mc@jvavav.com>
Date: Mon, 9 Dec 2024 07:37:53 +0800
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index b68770f6992b044d13e67d9affa6933c90ca8fb8..9698b27a1bf10d585428c164d5fd060b6200dbe2 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -26,6 +26,21 @@ public interface ServerBuildInfo {
     Key BRAND_GALE_ID = Key.key("galemc", "gale");
     // Gale end - branding changes
 
+    // Leaf start
+    /**
+     * The brand id for Purpur.
+     */
+    Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
+    /**
+     * The brand id for Leaf.
+     */
+    Key BRAND_LEAF_ID = Key.key("winds-studio", "leaf");
+    /**
+     * The brand id for Pufferfish.
+     */
+    Key BRAND_PUFFERFISH_ID = Key.key("pufferfish", "pufferfish");
+    // Leaf end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index a6a283f5cda987fdee4135186c2e865da2b0efc2..8f068dfe21544d9037ba1e45e8a2f0676947f3bc 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -215,7 +215,7 @@ public class VersionCommand extends BukkitCommand {
         String version = Bukkit.getVersion();
         // Paper start
         if (version.startsWith("null")) { // running from ide?
-            setVersionMessage(Component.text("Unknown version, custom build?", NamedTextColor.YELLOW));
+            setVersionMessage(Component.text("* Unknown version, custom build?", NamedTextColor.RED)); // Leaf - Purpur - Clean up version command output
             return;
         }
         setVersionMessage(getVersionFetcher().getVersionMessage(version));
